starknet_devnet/cairo_rs_py_patch.py

```diff
@@ -0,0 +1,577 @@
+        index=segment_base_ptr.segment_index
+    )
+
+    stark_assert(
+        expected_stop_ptr == segment_stop_ptr,
+        code=StarknetErrorCode.SECURITY_ERROR,
+        message=(
+            f"Invalid stop pointer for segment. "
+            f"Expected: {expected_stop_ptr}, found: {segment_stop_ptr}."
+        ),
+    )
+
+
+def cairo_rs_py_get_return_values(runner: CairoFunctionRunner) -> List[int]:
+    """
+    Extracts the return values of a StarkNet contract function from the Cairo runner.
+    """
+    with wrap_with_stark_exception(
+        code=StarknetErrorCode.INVALID_RETURN_DATA,
+        message="Error extracting return data.",
+        logger=logger,
+        exception_types=[Exception],
+    ):
+        ret_data_size, ret_data_ptr = runner.get_return_values(2)
+        values = runner.memory.get_range(ret_data_ptr, ret_data_size)
+        values = runner.get_range(ret_data_ptr, ret_data_size)
+        stark_assert(
+            all(isinstance(value, int) for value in values),
+            code=StarknetErrorCode.INVALID_RETURN_DATA,
+            message="Return data expected to be non-relocatable.",
+        )
+    return cast(List[int], values)
+
+
+def cairo_rs_py_validate_read_only_segments(self, runner: CairoFunctionRunner):
+    """
+    Validates that there were no out of bounds writes to read-only segments and marks
+    them as accessed.
+    """
+    segments = runner
+
+    for segment_ptr, segment_size in self.read_only_segments:
+        used_size = segments.get_segment_used_size(index=segment_ptr.segment_index)
+        stark_assert(
+            used_size == segment_size,
+            code=StarknetErrorCode.SECURITY_ERROR,
+            message="Out of bounds write to a read-only segment.",
+        )
+        runner.mark_as_accessed(address=segment_ptr, size=segment_size)
+
+
+def cairo_rs_py_monkeypatch():
+    setattr(ExecuteEntryPoint, "_run", cairo_rs_py_run)
+    setattr(
+        sys.modules["starkware.starknet.core.os.class_hash"],
+        "class_hash_inner",
+        cairo_rs_py_compute_class_hash_inner,
+    )
+    setattr(
+        sys.modules["starkware.starknet.core.os.os_utils"],
+        "prepare_os_context",
+        cairo_rs_py_prepare_os_context,
+    )
+    setattr(
+        sys.modules["starkware.starknet.core.os.os_utils"],
+        "validate_and_process_os_context",
+        cairo_rs_py_validate_and_process_os_context,
+    )
+    setattr(
+        BusinessLogicSysCallHandler, "_allocate_segment", cairo_rs_py_allocate_segment
+    )
+    setattr(
+        BusinessLogicSysCallHandler,
+        "_read_and_validate_syscall_request",
+        cairo_rs_py_read_and_validate_syscall_request,
+    )
+    setattr(
+        BusinessLogicSysCallHandler,
+        "validate_read_only_segments",
+        cairo_rs_py_validate_read_only_segments,
+    )
+    setattr(
+        sys.modules["starkware.starknet.business_logic.transaction.fee"],
+        "calculate_l1_gas_by_cairo_usage)",
```

Comment:
Is this causing `test/rpc/test_rpc_estimate_fee.py::test_estimate_fee_with_invalid_call_data` to fail?